File: Leet 3568/solution.ts
type State = {
  row: number;
  col: number;
  mask: number;
  remainingEnergy: number;
  steps: number;
};

type Position = {
  x: number;
  y: number;
}

function findStart(classroom: string[]): Position | null {
  for (let i = 0; i < classroom.length; i++) {
    let idx = classroom[i].indexOf("S");
    if (idx !== -1) {
      return { x: i, y: idx };
    }
  }
  return null;
}

function minMoves(classroom: string[], energy: number): number {
  if (classroom.length === 0 || classroom[0].length === 0) {
    return -1;
  }

  const startPos = findStart(classroom);

  if (startPos === null) {
    return -1;
  }

  const rows = classroom.length;
  const cols = classroom[0].length;

  // store classroom litter locations along with id for mask
  const litter = new Map<number, number>();
  let id = 0;
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      if (classroom[row][col] === "L") {
        litter.set(row * cols + col, id++);
      }
    }
  }

  let target = (1 << litter.size) - 1;

  let best: number[][][] = Array.from({ length: target + 1 }, () =>
    Array.from({ length: rows }, () => Array(cols).fill(-1)),
  );
  
  const directions: Position[] = [
    { x: 0, y: 1 },
    { x: 0, y: -1 },
    { x: 1, y: 0 },
    { x: -1, y: 0 },
  ];

  let frame: State[] = [{ row: startPos.x, col: startPos.y, mask: 0, remainingEnergy: energy, steps: 0 }];
  while (frame.length > 0) {
    let { row, col, mask, remainingEnergy, steps } = frame.shift()!;

    // reset energy
    if (classroom[row][col] === "R") {
      remainingEnergy = energy;
    }

    // let litterID = litter.get(mapKey(row, col));
    let litterId = litter.get(row * cols + col);
    if (litterId !== undefined) {
      mask |= 1 << litterId;
    }

    // all litter picked up
    if (mask === target) {
      return steps;
    }

    // already visited
    if (remainingEnergy <= best[mask][row][col]) {
      continue;
    }

    best[mask][row][col] = remainingEnergy;

    // cant proceed any further as energy is needed to move
    if (remainingEnergy === 0) {
      continue;
    }

    // explore all directions
    for (let { x, y } of directions) {
      let nextRow = row + x;
      let nextCol = col + y;

      // out of bounds
      if (nextRow < 0 || nextCol < 0 || nextRow >= rows || nextCol >= cols) {
        continue;
      }

      // blocked
      if (classroom[nextRow][nextCol] === "X") {
        continue;
      }

      frame.push({
        row: nextRow,
        col: nextCol,
        mask: mask,
        remainingEnergy: remainingEnergy - 1,
        steps: steps + 1,
      });
    }
  }

  // no solution found
  return -1;
}

function main(): void {
  console.log("3568. Minimum Moves to Clean the Classroom");

  let result: number = minMoves(["S.", "XL"], 2);
  console.log(`Test case for [S., XL] should equal 2 ? ${result}`);
}

main();

File: Leet 3568/solution.cs
class Solution
{
    private static (int x, int y) FindStart(string[] classroom)
    {
        for(int i = 0; i < classroom.Length; i++)
        {
            var j = classroom[i].IndexOf('S');
            if (j != -1)
            {
                return (i, j);
            }
        }

        return (-1, -1);
    }

    public static int MinMoves(string[] classroom, int energy)
    {
        if (classroom.Length == 0 || classroom[0].Length == 0)
        {
            return -1;
        }

        var (x, y) = FindStart(classroom);
        if (x == -1 || y == -1)
        {
            return -1;
        }

        var rows = classroom.Length;
        var cols = classroom[0].Length;

        var litter = new Dictionary<int, int>();
        var id = 0;
        for (var r = 0; r < rows; r++)
        {
            for (var c = 0; c < cols; c++)
            {
                if (classroom[r][c] == 'L')
                {
                    litter.Add(r * cols + c, id++);
                }
            }
        }

        var target = (1 << litter.Count) - 1;

        var best = new int[target + 1, rows, cols];
        for (var i = 0; i < best.GetLength(0); i++)
        {
            for (var j = 0; j < best.GetLength(1); j++)
            {
                for (var k = 0; k < best.GetLength(2); k++)
                {
                    best[i, j, k] = -1;
                }
            }
        }

        var directions = new[] { (0, 1), (0, -1), (1, 0), (-1, 0) };

        var queue = new Queue<(int row, int col, int mask, int remainingEnergy, int steps)>();
        queue.Enqueue((x, y, 0, energy, 0));

        while (queue.Count > 0)
        {
            var (row, col, mask, remainingEnergy, steps) = queue.Dequeue();

            if (classroom[row][col] == 'R')
            {
                remainingEnergy = energy;
            }

            if (litter.TryGetValue(row * cols + col, out var litterId))
            {
                mask |= 1 << litterId;
            }

            if (mask == target)
            {
                return steps;
            }

            if (best[mask, row, col] >= remainingEnergy)
            {
                continue;
            }

            best[mask, row, col] = remainingEnergy;

            if (remainingEnergy == 0)
            {
                continue;
            }

            foreach (var (dx, dy) in directions)
            {
                var nextRow = row + dx;
                var nextCol = col + dy;

                if (nextRow < 0 || nextCol < 0 || nextRow >= rows || nextCol >= cols)
                {
                    continue;
                }

                if (classroom[nextRow][nextCol] == 'X')
                {
                    continue;
                }

                queue.Enqueue((nextRow, nextCol, mask, remainingEnergy - 1, steps + 1));
            }
        }

        return -1;
    }
}

class Program
{
    private static void Main()
    {
        Console.WriteLine("3568. Minimum Moves to Clean the Classroom");

        int result = Solution.MinMoves(["S.", "XL"], 2);
        Console.WriteLine($"Result: {result}");
    }
}